File: frontend/src/utils/validators.js
/**
 * utils/validators.js
 * ────────────────────
 * Zod schemas for form validation — import and reuse across React Hook Form.
 */

import { z } from "zod";

// ── Common field schemas ───────────────────────────────────────────────────────
export const emailSchema = z
  .string()
  .min(1, "Email is required")
  .email("Enter a valid email address");

export const passwordSchema = z
  .string()
  .min(8, "Password must be at least 8 characters")
  .regex(/[A-Z]/, "Must contain at least one uppercase letter")
  .regex(/[0-9]/, "Must contain at least one number");

export const phoneSchema = z
  .string()
  .regex(/^\+?[1-9]\d{6,14}$/, "Enter a valid phone number")
  .optional();

// ── Auth schemas ───────────────────────────────────────────────────────────────
export const loginSchema = z.object({
  email: emailSchema,
  password: z.string().min(1, "Password is required"),
});

export const registerSchema = z
  .object({
    name: z.string().min(2, "Name must be at least 2 characters"),
    email: emailSchema,
    password: passwordSchema,
    confirmPassword: z.string(),
  })
  .refine((data) => data.password === data.confirmPassword, {
    message: "Passwords do not match",
    path: ["confirmPassword"],
  });

export const forgotPasswordSchema = z.object({
  email: emailSchema,
});

File: backend/app/utils/validators.py
"""
app/utils/validators.py
────────────────────────
Reusable field validators for Pydantic schemas.
"""

import re


EMAIL_REGEX = re.compile(r"^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$")
PHONE_REGEX = re.compile(r"^\+?[1-9]\d{6,14}$")


def validate_email_format(email: str) -> str:
    if not EMAIL_REGEX.match(email):
        raise ValueError(f"Invalid email address: {email}")
    return email.lower()


def validate_phone_format(phone: str) -> str:
    if not PHONE_REGEX.match(phone):
        raise ValueError(f"Invalid phone number: {phone}")
    return phone


def validate_password_strength(password: str) -> str:
    """
    Enforce minimum password requirements:
    - At least 8 characters
    - At least one uppercase letter
    - At least one digit
    """
    if len(password) < 8:
        raise ValueError("Password must be at least 8 characters long.")
    if not any(c.isupper() for c in password):
        raise ValueError("Password must contain at least one uppercase letter.")
    if not any(c.isdigit() for c in password):
        raise ValueError("Password must contain at least one digit.")
    return password
